drivers/soundwire/intel_init.c

```diff
@@ -32,30 +32,73 @@ static void intel_link_auxdev_release(struct device *dev)
 	kfree(ldev);
 }
 
-static int intel_link_dev_init(struct sdw_intel_link_dev *ldev,
-			       struct device *parent,
-			       struct fwnode_handle *fwnode,
-			       const char *name,
-			       int link_id)
+/* alloc, init and add link devices */
+static struct sdw_intel_link_dev *intel_link_dev_register(struct sdw_intel_res *res,
+							  struct sdw_intel_ctx *ctx,
+							  struct fwnode_handle *fwnode,
+							  const char *name,
+							  int link_id)
 {
+	struct sdw_intel_link_dev *ldev;
+	struct sdw_intel_link_res *link;
 	struct auxiliary_device *auxdev;
 	int ret;
 
+	ldev = kzalloc(sizeof(*ldev), GFP_KERNEL);
+	if (!ldev)
+		return ERR_PTR(-ENOMEM);
+
 	auxdev = &ldev->auxdev;
 	auxdev->name = name;
-	auxdev->dev.parent = parent;
+	auxdev->dev.parent = res->parent;
 	auxdev->dev.fwnode = fwnode;
 	auxdev->dev.release = intel_link_auxdev_release;
 
 	/* we don't use an IDA since we already have a link ID */
 	auxdev->id = link_id;
 
 	ret = auxiliary_device_init(auxdev);
-	if (ret < 0)
-		dev_err(parent, "failed to initialize link dev %s link_id %d\n",
+	if (ret < 0) {
+		dev_err(res->parent, "failed to initialize link dev %s link_id %d\n",
 			name, link_id);
+		kfree(ldev);
+		return ERR_PTR(ret);
+	}
+
+	/*
+	 * keep a handle on the allocated memory, to be used in all other functions.
+	 * Since the same pattern is used to skip links that are not enabled, there is
+	 * not need to check if ctx->ldev[i] is NULL later on.
+	 */
+	ctx->ldev[link_id] = ldev;
+
+	/* now initialize link information */
+	link = &ldev->link_res;
+	link->mmio_base = res->mmio_base;
+	link->registers = res->mmio_base + SDW_LINK_BASE
+		+ (SDW_LINK_SIZE * link_id);
+	link->shim = res->mmio_base + SDW_SHIM_BASE;
+	link->alh = res->mmio_base + SDW_ALH_BASE;
+
+	link->ops = res->ops;
+	link->dev = res->dev;
+
+	link->clock_stop_quirks = res->clock_stop_quirks;
+	link->shim_lock = &ctx->shim_lock;
+	link->shim_mask = &ctx->shim_mask;
+	link->link_mask = ctx->link_mask;
+
+	/* everything initialized, last step is to add device */
+	ret = auxiliary_device_add(&ldev->auxdev);
+	if (ret < 0) {
+		dev_err(res->parent, "failed to add link dev %s link_id %d\n",
+			ldev->auxdev.name, link_id);
+		/* ldev will be freed with the put_device() and .release sequence */
+		auxiliary_device_uninit(&ldev->auxdev);
+		return ERR_PTR(ret);
+	}
```

Comment:
I think its better to keep the device_add() right after device_init() in this function. I dont see anything that can fail between lines 67-91 anyway.

Reply:
no, we can't do this, these inits are required, they will be used in the driver probe.

Reply:
I see how about the other way then? Move the device_init down to be just before device_add()? I just feel that if we split it, it will create openings for the same kind of questions. and in the future, we someone successfully manages to modify the API to unify init and add, the change will be simpler for us

Reply:
that would probably work, yes, good suggestion @ranj063 
so the steps would be alloc, initialize structure, device_init, device_add.

Reply:
done